File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/TipoPropostaService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.TipoProposta;
import br.com.portalproposta.backend.repository.PropostaRepository;
import br.com.portalproposta.backend.repository.TipoPropostaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class TipoPropostaService {

    private final TipoPropostaRepository tipoPropostaRepository;
    private final PropostaRepository propostaRepository;

    @Autowired
    public TipoPropostaService(TipoPropostaRepository tipoPropostaRepository, PropostaRepository propostaRepository) {
        this.tipoPropostaRepository = tipoPropostaRepository;
        this.propostaRepository = propostaRepository;
    }

    public List<TipoProposta> list() {
        return tipoPropostaRepository.findAllByOrderByNomeAsc();
    }

    public TipoProposta getById(String id) {
        return tipoPropostaRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("tipo de proposta não encontrado"));
    }

    @Transactional
    public TipoProposta create(TipoProposta tipo) {
        if (tipo.getNome() == null || tipo.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("o nome do tipo de proposta é obrigatório");
        }

        if (tipo.getChave() == null || tipo.getChave().trim().isEmpty()) {
            throw new AppBadRequestException("a chave do tipo de proposta é obrigatória");
        }

        String cleanedChave = tipo.getChave().replaceAll("[^a-zA-Z0-9]", "");
        if (cleanedChave.isEmpty()) {
            throw new AppBadRequestException("a chave do tipo de proposta deve conter caracteres alfanuméricos");
        }
        tipo.setChave(cleanedChave);

        tipoPropostaRepository.findByChave(tipo.getChave()).ifPresent(existing -> {
            throw new AppBadRequestException("já existe um tipo de proposta com esta chave");
        });

        tipo.setId(UUID.randomUUID().toString());
        tipo.setCreatedAt(LocalDateTime.now());
        tipo.setUpdatedAt(LocalDateTime.now());

        return tipoPropostaRepository.save(tipo);
    }

    @Transactional
    public TipoProposta update(String id, TipoProposta input) {
        TipoProposta tipo = getById(id);

        if (input.getNome() == null || input.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("o nome do tipo de proposta é obrigatório");
        }

        boolean isLegacy = tipo.getChave().equals("Imobiliaria")
                || tipo.getChave().equals("Auto")
                || tipo.getChave().equals("Comissionados");

        if (!isLegacy && input.getChave() != null && !input.getChave().trim().isEmpty()
                && !input.getChave().equals(tipo.getChave())) {
            String newChave = input.getChave().replaceAll("[^a-zA-Z0-9]", "");
            if (newChave.isEmpty()) {
                throw new AppBadRequestException("a chave do tipo de proposta deve conter caracteres alfanuméricos");
            }

            tipoPropostaRepository.findByChave(newChave).ifPresent(existing -> {
                if (!existing.getId().equals(tipo.getId())) {
                    throw new AppBadRequestException("já existe um tipo de proposta com esta chave");
                }
            });
            tipo.setChave(newChave);
        }

        tipo.setNome(input.getNome());
        tipo.setCampos(input.getCampos());
        tipo.setUpdatedAt(LocalDateTime.now());

        return tipoPropostaRepository.save(tipo);
    }

    @Transactional
    public void delete(String id) {
        TipoProposta tipo = getById(id);

        boolean isLegacy = tipo.getChave().equals("Imobiliaria")
                || tipo.getChave().equals("Auto")
                || tipo.getChave().equals("Comissionados");

        if (isLegacy) {
            throw new AppBadRequestException("não é permitido excluir tipos de proposta do sistema");
        }

        long count = propostaRepository.countByTipo(tipo.getChave());
        if (count > 0) {
            throw new AppBadRequestException(String.format(
                    "não é possível excluir este tipo de proposta pois existem %d proposta(s) associada(s) a ele",
                    count));
        }

        tipoPropostaRepository.delete(tipo);
    }
}

File: backend-net10/Services/TipoPropostaService.cs
using Microsoft.EntityFrameworkCore;
using backend_net10.Data;
using backend_net10.Models;
using System.Text.RegularExpressions;

namespace backend_net10.Services;

public class TipoPropostaService
{
    private readonly AppDbContext _context;

    public TipoPropostaService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<TipoProposta>> ListAsync()
    {
        return await _context.TiposProposta.OrderBy(t => t.Nome).ToListAsync();
    }

    public async Task<TipoProposta> GetByIdAsync(string id)
    {
        var type = await _context.TiposProposta.FirstOrDefaultAsync(t => t.Id == id);
        if (type == null)
        {
            throw new AppNotFoundException("tipo de proposta não encontrado");
        }
        return type;
    }

    public async Task<TipoProposta> CreateAsync(TipoProposta tipo)
    {
        if (string.IsNullOrWhiteSpace(tipo.Nome))
        {
            throw new AppBadRequestException("o nome do tipo de proposta é obrigatório");
        }

        if (string.IsNullOrWhiteSpace(tipo.Chave))
        {
            throw new AppBadRequestException("a chave do tipo de proposta é obrigatória");
        }

        tipo.Chave = Regex.Replace(tipo.Chave, @"[^a-zA-Z0-9]", "");
        if (string.IsNullOrEmpty(tipo.Chave))
        {
            throw new AppBadRequestException("a chave do tipo de proposta deve conter caracteres alfanuméricos");
        }

        var existing = await _context.TiposProposta.FirstOrDefaultAsync(t => t.Chave == tipo.Chave);
        if (existing != null)
        {
            throw new AppBadRequestException("já existe um tipo de proposta com esta chave");
        }

        tipo.Id = Guid.NewGuid().ToString();
        tipo.CreatedAt = DateTime.Now;
        tipo.UpdatedAt = DateTime.Now;

        _context.TiposProposta.Add(tipo);
        await _context.SaveChangesAsync();

        return tipo;
    }

    public async Task<TipoProposta> UpdateAsync(string id, TipoProposta input)
    {
        var tipo = await _context.TiposProposta.FirstOrDefaultAsync(t => t.Id == id);
        if (tipo == null)
        {
            throw new AppNotFoundException("tipo de proposta não encontrado");
        }

        if (string.IsNullOrWhiteSpace(input.Nome))
        {
            throw new AppBadRequestException("o nome do tipo de proposta é obrigatório");
        }

        bool isLegacy = tipo.Chave == "Imobiliaria" || tipo.Chave == "Auto" || tipo.Chave == "Comissionados";

        if (!isLegacy && !string.IsNullOrWhiteSpace(input.Chave) && input.Chave != tipo.Chave)
        {
            var newChave = Regex.Replace(input.Chave, @"[^a-zA-Z0-9]", "");
            if (string.IsNullOrEmpty(newChave))
            {
                throw new AppBadRequestException("a chave do tipo de proposta deve conter caracteres alfanuméricos");
            }

            var existing = await _context.TiposProposta.FirstOrDefaultAsync(t => t.Chave == newChave);
            if (existing != null && existing.Id != tipo.Id)
            {
                throw new AppBadRequestException("já existe um tipo de proposta com esta chave");
            }
            tipo.Chave = newChave;
        }

        tipo.Nome = input.Nome;
        tipo.Campos = input.Campos;
        tipo.UpdatedAt = DateTime.Now;

        await _context.SaveChangesAsync();
        return tipo;
    }

    public async Task DeleteAsync(string id)
    {
        var tipo = await _context.TiposProposta.FirstOrDefaultAsync(t => t.Id == id);
        if (tipo == null)
        {
            throw new AppNotFoundException("tipo de proposta não encontrado");
        }

        if (tipo.Chave == "Imobiliaria" || tipo.Chave == "Auto" || tipo.Chave == "Comissionados")
        {
            throw new AppBadRequestException("não é permitido excluir tipos de proposta do sistema");
        }

        var count = await _context.Propostas.CountAsync(p => p.Tipo == tipo.Chave);
        if (count > 0)
        {
            throw new AppBadRequestException($"não é possível excluir este tipo de proposta pois existem {count} proposta(s) associada(s) a ele");
        }

        _context.TiposProposta.Remove(tipo);
        await _context.SaveChangesAsync();
    }
}
